R/downloader.R

```diff
@@ -244,28 +282,85 @@ secureDownloadMethod <- function() {
     # that doesn't rely on the value of setInternet2). If it's R <= 3.1
     # then we can use "internal" for https so long as internet2 is enabled
     # (we don't use libcurl on Windows because it doesn't check certs).
-    if (isR32)
-        return("wininet")
-
-    # Otherwise, make a call to 'setInternet2' and use the 'internal' method
-    # if that call succeeds.
-    seti2 <- `::`(utils, 'setInternet2')
-    if (suppressWarnings(seti2(NA)))
-      return("internal")
+    if (isR32) {
+      return("wininet")
+    } else {
+      # Otherwise, make a call to 'setInternet2' and use the 'internal' method
+      # if that call succeeds.
+      seti2 <- `::`(utils, 'setInternet2')
+      if (suppressWarnings(seti2(NA))) {
+        return("internal")
+      }
+    }
   }
 
-  # For Darwin and Linux we use libcurl if we can and then fall back
-  # to curl or wget as appropriate. We prefer libcurl because it honors
-  # the same proxy configuration that "internal" does so it less likely
-  # to break downloads for users behind proxy servers.
-  if (canUseLibCurlDownloadMethod())
-    return("libcurl")
-
-  # Otherwise, fall back to 'wget' or 'curl' (preferring 'wget')
+  # Otherwise, fall back to 'wget' or 'curl' (preferring 'curl')
   candidates <- c("curl", "wget")
   for (candidate in candidates)
     if (isProgramOnPath(candidate))
       return(candidate)
 
   stop("Failed to discover a secure download method.")
 }
+
+# Returns a function to be used as an error handler for githubDownload,
+# gitlabDownload, and bitbucketDownload.
+authDownloadAdvice <- function(type = NULL) {
+  f <- function(e = NULL) {
+    advice <- paste0(
+      "If you are trying to restore a package from a private Git repo, ",
+      "you must have credentials available in your environment, ",
+      "and Packrat must be configured to use an auth-capable download method."
+    )
+
+    # Info on available auth tokens, dependant on provider type.
+    token_msg <- NULL
+    if (identical(type, "github")) {
+      if (githubAuthenticated()) {
+        token_msg <- "GITHUB_PAT found; check that it is correct."
+      } else {
+        token_msg <- "GITHUB_PAT environment variable not found."
+      }
+    }
+    if (identical(type, "gitlab")) {
+      if (gitlabAuthenticated()) {
+        token_msg <- "GITLAB_PAT found; check that it is correct."
+      } else {
+        token_msg <- "GITLAB_PAT environment variable not found."
+      }
+    }
+    if (identical(type, "bitbucket")) {
+      if (bitbucketAuthenticated()) {
+        token_msg <- "BITBUCKET_USERNAME and BITBUCKET_PASSWORD found; check that they are correct."
+      } else {
+        token_msg <- "BITBUCKET_USERNAME and BITBUCKET_PASSWORD environment variables not found."
+      }
+    }
+    advice <- c(advice, token_msg)
+
+    # Info on configuration
+    if (canUseRenvDownload()) {
```

Comment:
This control logic feels repetitive; when generating this advice, we know if authentication is in use and what provider was used. Those can all be inputs to this function.

```r
authDownloadAdvice <- function(type, authenticated, downloader) { ... }
```

```r
tryCatch(renvDownload(url, destfile, type = "github"), error = authDownloadAdvice("github", TRUE, "renv")
```

If we pass information about what was used, this function is still useful even if we change the conditions we use to decide which technique is in play.

Reply:
This makes a lot of sense, and… it *may* be possible to do it and keep the logic of the download function selection readable.

Reply:
Done.